prune/prune.go

```diff
@@ -16,184 +16,209 @@ package prune
+// Pruner is an object that runs a prune job.
+type Pruner struct {
+	Registry
+
+	// Client is the controller-runtime client that will be used
+	// To perform a dry run, use the controller-runtime DryRunClient
+	Client client.Client
+
+	// GVK is the type of objects to prune.
+	// It defaults to Pod
+	GVK schema.GroupVersionKind
+
+	// Strategy is the function used to determine a list of resources that are pruneable
+	Strategy StrategyFunc
+
+	// Labels is a map of the labels to use for label matching when looking for resources
+	Labels map[string]string
+
+	// Namespace is the namespace to use when looking for resources
+	Namespace string
+
+	// Logger is the logger to use when running pruning functionality
+	Logger logr.Logger
 }
 
-// Execute causes the pruning work to be executed based on its configuration
-func (config Config) Execute(ctx context.Context) error {
-	log := Logger(ctx, config)
-	log.V(1).Info("Execute Prune")
+// Unprunable indicates that it is not allowed to prune a specific object.
+type Unprunable struct {
+	Obj    *client.Object
+	Reason string
+}
 
-	err := config.validate()
-	if err != nil {
-		return err
-	}
-
-	for i := 0; i < len(config.Resources); i++ {
-		var resourceList []ResourceInfo
-		var err error
-
-		if config.Resources[i].Kind == PodKind {
-			resourceList, err = config.getSucceededPods(ctx)
-			if err != nil {
-				return err
-			}
-			log.V(1).Info("pods ", "count", len(resourceList))
-		} else if config.Resources[i].Kind == JobKind {
-			resourceList, err = config.getCompletedJobs(ctx)
-			if err != nil {
-				return err
-			}
-			log.V(1).Info("jobs ", "count", len(resourceList))
-		}
+// Error returns a string representation of an `Unprunable` error.
+func (e *Unprunable) Error() string {
+	return fmt.Sprintf("unable to prune %s: %s", client.ObjectKeyFromObject(*e.Obj), e.Reason)
+}
 
-		var resourcesToRemove []ResourceInfo
-
-		switch config.Strategy.Mode {
-		case MaxAgeStrategy:
-			resourcesToRemove, err = pruneByMaxAge(ctx, config, resourceList)
-		case MaxCountStrategy:
-			resourcesToRemove, err = pruneByMaxCount(ctx, config, resourceList)
-		case CustomStrategy:
-			resourcesToRemove, err = config.CustomStrategy(ctx, config, resourceList)
-		default:
-			return fmt.Errorf("unknown strategy")
-		}
-		if err != nil {
-			return err
-		}
+// StrategyFunc takes a list of resources and returns the subset to prune.
+type StrategyFunc func(ctx context.Context, objs []client.Object) ([]client.Object, error)
 
-		err = config.removeResources(ctx, resourcesToRemove)
-		if err != nil {
-			return err
-		}
-	}
+// IsPrunableFunc is a function that checks the data of an object to see whether or not it is safe to prune it.
+// It should return `nil` if it is safe to prune, `Unprunable` if it is unsafe, or another error.
+// It should safely assert the object is the expected type, otherwise it might panic.
+type IsPrunableFunc func(obj client.Object) error
```

Comment:
What do you think of passing the logger  to the function?

Reply:
I'm not sure that is necessary since we are planning on using the logger in the `Pruner`'s `Prune` function to log when an object is not prunable. 

Is there a use case that you have in mind that might require the logger being passed into an `IsPrunableFunc`?

Reply:
My thinking is that IsPrunableFunc will usually be implemented by the library users. It is difficult to anticipate what logic will be built in there and whether they want to log states that are noteworthy.

Reply:
That is a fair point, I don't think it hurts to pass it so might as well. Will work on making this change.

Reply:
done in https://github.com/operator-framework/operator-lib/pull/105/commits/67688741005b777c097f0c3952c3d6c010d15db9